gui/widgets/grid.cpp

```diff
@@ -0,0 +1,640 @@
+#include "gui/widgets/grid.h"
+
+#include "gui/ThemeEval.h"
+
+namespace GUI {
+
+GridItemWidget::GridItemWidget(GridWidget *boss, int x, int y, int w, int h)
+	: ContainerWidget(boss, x, y, w, h), CommandSender(boss) {
+
+	setFlags(WIDGET_ENABLED | WIDGET_TRACK_MOUSE | WIDGET_CLEARBG);
+	_activeEntry = nullptr;
+	_grid = boss;
+	isHighlighted = false;
+}
+GridItemWidget::GridItemWidget(GridWidget *boss)
+	: GridItemWidget(boss, 0, 0, 0, 0) {}
+
+void GridItemWidget::setActiveEntry(GridItemInfo &entry) {
+	_activeEntry = &entry;
+}
+
+void GridItemWidget::updateThumb() {
+	const Graphics::ManagedSurface *gfx = _grid->filenameToSurface(_activeEntry->thumbPath);
+	_thumbGfx.free();
+	if (gfx)
+		_thumbGfx.copyFrom(*gfx);
+}
+
+void GridItemWidget::update() {
+	if (_activeEntry) {
+		updateThumb();
+		setTooltip(_activeEntry->title);
+		markAsDirty();
+	}
+}
+
+void GridItemWidget::move(int x, int y) {
+	Widget::setPos(getRelX() + x, getRelY() + y);
+}
+
+void GridItemWidget::drawWidget() {
+	int thumbHeight = _grid->getThumbnailHeight();
+	int thumbWidth = _grid->getThumbnailWidth();
+	Array<U32String> titleLines;
+	g_gui.getFont().wordWrapText(_activeEntry->title, thumbWidth, titleLines);
+
+	if ((isHighlighted) || (_grid->getSelected() == _activeEntry->entryID)) {
+		// Draw a highlighted BG on hover
+		g_gui.theme()->drawWidgetBackground(Common::Rect(_x - (_grid->_gridXSpacing / 3), _y - (_grid->_gridYSpacing / 3),
+											_x + _w + (_grid->_gridXSpacing / 3), _y + _h + (_grid->_gridYSpacing / 3)), 
+										ThemeEngine::WidgetBackground::kGridItemHighlight);
+	} else {
+		// Draw a BG of the same color as grid area
+		// when it is not highlighted to cover up 
+		// the highlight shadow
+		// FIXME: Find a way to redraw the area around the widget
+		//		 instead of just drawing a cover-up
+		g_gui.theme()->drawWidgetBackground(Common::Rect(_x - 2 * (_grid->_gridXSpacing / 3), _y - 2 * (_grid->_gridYSpacing / 3),
+											_x + _w + 2 * (_grid->_gridXSpacing / 3), _y + _h + 2 * (_grid->_gridYSpacing / 3)), 
+										ThemeEngine::WidgetBackground::kGridItemBackground);
+	}
+
+	// Draw Thumbnail Background
+	g_gui.theme()->drawWidgetBackground(Common::Rect(_x, _y, _x + _grid->getThumbnailWidth(), _y + thumbHeight), 
+										ThemeEngine::WidgetBackground::kThumbnailBackground);
+	
+	// Draw Thumbnail
+	if (_thumbGfx.empty()) {
+		// Draw Title when thumbnail is missing
+		int linesInThumb = MIN(thumbHeight / kLineHeight, (int)titleLines.size());
+		for (int i = 0; i < linesInThumb; ++i) {
+			g_gui.theme()->drawText(Common::Rect(_x, _y + ((thumbHeight - (linesInThumb - 2 * i) * kLineHeight)) / 2,
+							_x + thumbWidth, _y + ((thumbHeight - (linesInThumb - 2 * i) * kLineHeight)) / 2 + kLineHeight),
+							titleLines[i], GUI::ThemeEngine::kStateEnabled, Graphics::kTextAlignCenter,
+							ThemeEngine::kTextInversionNone, 0, true, ThemeEngine::kFontStyleNormal,
+							ThemeEngine::kFontColorAlternate, false);
+		}
+	} else
```

Comment:
Braces, especially when the if has braces.

Reply:
Formatted with braces in 739cb8e.